app/__main__.py

```diff
@@ -0,0 +1,107 @@
+import abc
+import csv
+import datetime
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Union, Dict, List, Optional, Generator
+
+import dateutil
+import dateutil.parser
+
+
+@dataclass
+class Person:
+    address: str
+    data: Dict[str, Union[str, datetime.datetime]]
+
+
+class BaseResource(abc.ABC):
+
+    @abc.abstractmethod
+    def connect(self):
+        ...
+
+    @abc.abstractmethod
+    def parse(self):
+        ...
+
+
+class CSVResource(BaseResource):
+    def __init__(self, path):
+        self._path = path
+        self._reader: Optional[List[Dict[str, Union[str, datetime]]]] = None
+
+    def connect(self) -> None:
+        with open(self._path, mode='r') as csv_input:
+            self._reader = list(csv.DictReader(csv_input))
+
+    def parse(self) -> Generator:
+        assert self._reader
```

Comment:
absolutely agree, I am completely not fun of assertion, but this specific assertion is caused by _mypy_ - it throws error because by default `self._parser = None`. :) 

It can take a while to refactor mypy here, just decided to make a simplest fix.